File: Python/camera-web-stream.py
from picamera2 import Picamera2
from flask import Flask, Response
import cv2
import numpy as np
import threading
import time
import psutil

app = Flask(__name__)

# 高级摄像头配置
picam2 = Picamera2()
config = picam2.create_video_configuration(
    main={"size": (2592, 1944)},  # 原生分辨率
    lores={"size": (2592, 1944)},  # 预览流
    display="main",
    encode="main",
    controls={"FrameRate": 30},
)
picam2.configure(config)
picam2.start()

frame_lock = threading.Lock()
latest_frame = None


def adaptive_quality():
    """动态质量调整"""
    cpu = psutil.cpu_percent()
    net = psutil.net_io_counters().bytes_sent
    return 90 if cpu < 70 and net < 1e6 else 75


def capture_frames():
    global latest_frame
    while True:
        with frame_lock:
            frame = picam2.capture_array("main")  # 获取高分辨率流
            frame = cv2.cvtColor(frame, cv2.COLOR_RGB2BGR)

            # 锐化处理
            kernel = np.array([[-1, -1, -1], [-1, 9, -1], [-1, -1, -1]])
            frame = cv2.filter2D(frame, -1, kernel)

            quality = adaptive_quality()
            _, jpeg = cv2.imencode(
                ".jpg", frame, [int(cv2.IMWRITE_JPEG_QUALITY), quality]
            )
            latest_frame = jpeg.tobytes()
        time.sleep(0.03)


def generate_stream():
    """生成 MJPEG 流"""
    while True:
        if latest_frame:
            yield (
                b"--frame\r\n"
                b"Content-Type: image/jpeg\r\n\r\n" + latest_frame + b"\r\n\r\n"
            )


@app.route("/video_feed")
def video_feed():
    """视频流路由"""
    return Response(
        generate_stream(), mimetype="multipart/x-mixed-replace; boundary=frame"
    )


@app.route("/")
def index():
    """显示监控页面"""
    return """
    <html>
      <head>
        <title>树莓派摄像头实时监控</title>
        <style>
          body { margin: 0; background: #000; }
          img { 
            width: 100vw; 
            height: 100vh;
            object-fit: contain;
          }
        </style>
      </head>
      <body>
        <img src="/video_feed">
      </body>
    </html>
    """


if __name__ == "__main__":
    # 启动摄像头捕获线程
    capture_thread = threading.Thread(target=capture_frames, daemon=True)
    capture_thread.start()

    # 启动 Web 服务器
    app.run(host="0.0.0.0", port=5000, threaded=True)

File: Haskell/camera-web-stream.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Concurrent (threadDelay)
import Control.Monad (forever, unless)
import qualified Data.ByteString as B
import Data.ByteString.Builder (
    Builder,
    byteString,
    stringUtf8,
 )
import Network.HTTP.Types (
    hContentType,
    status200,
    status404,
 )
import Network.Wai
import Network.Wai.Handler.Warp (run)
import System.IO (BufferMode (..), hSetBuffering)
import System.Process (
    StdStream (..),
    createPipe,
    createProcess,
    proc,
    std_out,
    waitForProcess,
 )

-- 配置参数
captureWidth :: String
captureWidth = "640"

captureHeight :: String
captureHeight = "480"

port :: Int
port = 8080

-- HTML 页面内容
htmlPage :: Builder
htmlPage =
    stringUtf8 $
        unlines
            [ "<!DOCTYPE html>"
            , "<html>"
            , "<head>"
            , "  <title>树莓派摄像头实时流</title>"
            , "  <meta charset=\"UTF-8\">"
            , "  <style>"
            , "    body { margin: 0; background: #000; }"
            , "    #video { width: 100%; height: 100vh; object-fit: contain; }"
            , "  </style>"
            , "</head>"
            , "<body>"
            , "  <img id=\"video\" src=\"/stream\">"
            , "</body>"
            , "</html>"
            ]

-- 摄像头进程初始化
startCameraProcess :: IO (IO B.ByteString)
startCameraProcess = do
    (_, Just hout, _, phandle) <-
        createProcess
            ( proc
                "libcamera-jpeg"
                [ "-o"
                , "-"
                , "--width"
                , captureWidth
                , "--height"
                , captureHeight
                , "--timelapse"
                , "33"
                , "--nopreview"
                , "--flush"
                , "--framerate"
                , "30"
                ]
            )
                { std_out = CreatePipe
                }

    hSetBuffering hout NoBuffering
    return $ B.hGetSome hout 8192

-- MJPEG 流响应
mjpegResponse :: IO Response
mjpegResponse = do
    getFrame <- startCameraProcess
    return
        $ responseStream
            status200
            [("Content-Type", "multipart/x-mixed-replace; boundary=frame")]
        $ \writeChunk _flush -> forever $ do
            frame <- getFrame
            unless (B.null frame) $ do
                writeChunk $
                    mconcat
                        [ stringUtf8 "--frame\r\n"
                        , stringUtf8 "Content-Type: image/jpeg\r\n\r\n"
                        , byteString frame
                        , stringUtf8 "\r\n"
                        ]

-- 应用路由
app :: Application
app req respond = case pathInfo req of
    [] -> respond $ responseBuilder status200 [htmlHeader] htmlPage
    ["stream"] -> do
        resp <- mjpegResponse
        respond resp
    _ -> respond $ responseBuilder status404 [] (stringUtf8 "Not Found")
  where
    htmlHeader = (hContentType, "text/html; charset=utf-8")

main :: IO ()
main = do
    putStrLn $ "服务已启动: http://0.0.0.0:" ++ show port
    run port app
